HeterogeneousCore/CUDATest/plugins/TestCUDAAnalyzerGPUKernel.h

```diff
@@ -0,0 +1,23 @@
+#ifndef HeterogeneousCore_CUDACore_TestCUDAAnalyzerGPUKernel_h
+#define HeterogeneousCore_CUDACore_TestCUDAAnalyzerGPUKernel_h
+
+#include "HeterogeneousCore/CUDAUtilities/interface/device_unique_ptr.h"
+
+#include <cuda_runtime.h>
+
+class TestCUDAAnalyzerGPUKernel {
+public:
+  static constexpr int NUM_VALUES = 4000;
+
+  TestCUDAAnalyzerGPUKernel(cudaStream_t stream);
+  ~TestCUDAAnalyzerGPUKernel() = default;
+
+  // returns (owning) pointer to device memory
+  void analyzeAsync(const float* d_input, cudaStream_t stream) const;
+  float value(cudaStream_t stream) const;
+
+private:
+  mutable cudautils::device::unique_ptr<float[]> sum_;  // all writes are atomic in CUDA
```

Comment:
Static analyzer complains about this `mutable`. All writes are in one kernel with `atomicAdd()`. Thinking out loud some options
1. Decorate the member with `CMS_SA_ALLOW`
   * Practical implication is that the very same branch could not be used anymore for 11_0_X backport as the `CMS_SA_ALLOW` does not exist there.
2. Technically the `mutable` is not needed here, because the `sum_` object is not modified, but `sum_.get()` and the `const` does not propagate there. The intention of the array pointed-to by `sum_` being modified in `analyseAsync()` becomes less clear.
3. The intention of `analyseAsync()` is to modify `sum_` so one could argue it should not be `const`. The function is `const` now because it is called from `edm::global::EDAnalyze::analyze()`, and the analyzer is `global` because "it can be" thanks to all writes being atomic. Removing `const` here would mean either adding `const_cast` in the `analyze()`, or changing the analyzer to `edm::one` (`edm::stream`+global cache would have the same `const` tension as the `edm::global` module).

Reply:
Discussed briefly with @Dr15Jones, we decided to go with 1 (mark with `CMS_THREAD_SAFE`).

Reply:
how would we handle it if it were an `std::vector<std::atomic<float>>` ?

Reply:
Another round of discussion, mainly around the constness of functions that mutate the visible state of the class, even if the mutation would be thread safe ("atomic histogram"). Perhaps the clearest would be to remove `mutable` here and make `analyzeAsync()` non-const, and mark the `TestCUDAAnalyzerGPU::gpuAlgo_` as `mutable`. Argumentation would be that filling the histogram does change the visible state of `TestCUDAAnalyzerGPUKernel`, but not really the visible state of `TestCUDAAnalyzerGPU`.